### PR TITLE
feat(js): add RouteConfigContext for path prefix customization [DRC-2334]

### PR DESCRIPTION
**Pull request by kentwelcome**
## Summary

Add `RouteConfigContext` to support path prefixing for recce-cloud navigation. This allows components using `useAppLocation` to navigate with automatic path prefix resolution when `RouteConfigProvider` is used.

### Problem

When launching Recce Web OSS from recce-cloud, clicking the "Query" button in the Lineage View's Explore dropdown navigates to `/query` (the OSS path) instead of `/oss/<sessionId>/query` (the Recce Web path).

### Solution

Add a new `RouteConfigContext` that provides path resolution. When `RouteConfigProvider` is used with a `basePath`, all navigation through `useAppLocation`, `useAppLocationWithSearch`, and `useAppNavigation` will automatically prefix paths.

### Changes

- **`js/src/lib/hooks/RouteConfigContext.tsx`** - New context provider with:
  - `RouteConfigProvider` - Provider component accepting `basePath` prop
  - `useRouteConfig()` - Hook returning `basePath` and `resolvePath()` function
  - `useRouteConfigSafe()` - Safe hook that returns null if no provider

- **`js/src/lib/hooks/useAppRouter.ts`** - Modified to use `RouteConfigContext`:
  - `useAppLocation` - Now resolves paths through `RouteConfigContext`
  - `useAppLocationWithSearch` - Now resolves paths through `RouteConfigContext`
  - `useAppNavigation` - `push` and `replace` now resolve paths

### Usage in recce-cloud

```tsx
import { RouteConfigProvider } from "@/lib/hooks/RouteConfigContext";

<RouteConfigProvider basePath={`/oss/${sessionId}`}>
  <RecceContextProvider>
    {children}
  </RecceContextProvider>
</RouteConfigProvider>
```

### Behavior

**With RouteConfigProvider:**
- `setLocation("/query")` → navigates to `/oss/<sessionId>/query`
- `setLocation("/checks?id=123")` → navigates to `/oss/<sessionId>/checks?id=123`

**Without RouteConfigProvider (OSS mode):**
- Paths work as before without any prefix (backward compatible)

## Test plan

- [x] Type check passes (`pnpm type:check`)
- [x] All existing tests pass (911 tests)
- [ ] Manual test in recce-cloud with RouteConfigProvider

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by kentwelcome** [comment]
## Addressed all review comments

Thank you for the thorough review! I've addressed all the feedback:

### 1. Path splitting logic (4 comments at lines 69, 111, 222, 229)
**Issue:** Using `split("?")` doesn't handle edge cases correctly (multiple `?` characters, fragment identifiers)

**Fix:** Replaced `split("?")` with `indexOf` approach that:
- Splits only on the first `?` occurrence
- Properly handles fragment identifiers (`#section`)
- Preserves both query strings and fragments in the resolved path

### 2. Missing test coverage (line 78)
**Issue:** No tests for RouteConfigContext integration

**Fix:** Added comprehensive test file `RouteConfigContext.test.tsx` with 28 tests covering:
- Path prefixing behavior with RouteConfigProvider
- Backward compatibility without provider
- Query string and fragment preservation
- Edge cases (paths already containing basePath, multiple `?` characters)
- Various URL schemes (http, https, mailto, tel, ftp, data)

### 3. URL scheme detection too narrow (line 78 in RouteConfigContext.tsx)
**Issue:** Only HTTP/HTTPS URLs were excluded from prefixing

**Fix:** Changed regex from `/^https?:\/\//` to `/^[a-z][a-z0-9+.-]*:/i` to match any valid URL scheme (mailto:, tel:, ftp:, data:, etc.)

### 4. Pathname not stripped of basePath (line 80)
**Issue:** Components receive full prefixed pathname, breaking route matching

**Fix:** Added `stripBasePath` function to RouteConfigContext and applied it to all hooks that return pathname:
- `useAppLocation`
- `useAppLocationWithSearch`
- `useAppNavigation`
- `useAppRoute`

Now components receive logical paths (e.g., `/checks`) instead of physical paths (e.g., `/oss/abc123/checks`), maintaining backward compatibility.

File: js/src/lib/hooks/useAppRouter.ts [hunk 1]
```diff
@@ -38,31 +45,44 @@ interface NavigateOptions {
...
   // Navigation function compatible with Wouter's setLocation
+  // Automatically applies basePath prefix from RouteConfigContext
   const setLocation = useCallback(
     (to: string, options?: NavigateOptions) => {
+      // Separate path and query string for proper handling
+      const [pathPart, queryPart] = to.split("?");
+      const resolvedPath = resolvePath(pathPart);
+      const fullPath = queryPart
+        ? `${resolvedPath}?${queryPart}`
+        : resolvedPath;
```
[the remark this answers is not in the document]
Comment (kentwelcome) [reply js/src/lib/hooks/useAppRouter.ts]: Fixed! Now using `indexOf` to split only on the first `?` occurrence and properly handling fragment identifiers (`#`). The implementation separates path, query string, and hash before resolving the path with basePath prefix.

File: js/src/lib/hooks/useAppRouter.ts [hunk 2]
```diff
@@ -38,31 +45,44 @@ interface NavigateOptions {
...
-        router.replace(to, { scroll: options?.scroll ?? true });
+        router.replace(fullPath, { scroll: options?.scroll ?? true });
       } else {
-        router.push(to, { scroll: options?.scroll ?? true });
+        router.push(fullPath, { scroll: options?.scroll ?? true });
       }
     },
-    [router],
+    [router, resolvePath],
   );
```
[the remark this answers is not in the document]
Comment (kentwelcome) [reply js/src/lib/hooks/useAppRouter.ts:106]: Added comprehensive test file `RouteConfigContext.test.tsx` with 28 tests covering:
- Path prefixing behavior with RouteConfigProvider
- Backward compatibility without provider (default resolvePath behavior)
- Query string and fragment preservation with path prefixing
- Edge cases (paths already containing basePath, multiple `?` characters)
- Various URL schemes (mailto:, tel:, ftp:, data:)

File: js/src/lib/hooks/useAppRouter.ts [hunk 3]
```diff
@@ -77,23 +97,30 @@ export function useAppLocationWithSearch(): [
...
   const location = useMemo(() => {
     const search = searchParams?.toString();
     return search ? `${pathname}?${search}` : pathname;
   }, [pathname, searchParams]);
 
-  // Navigation function compatible with Wouter's setLocation
+  // Navigation function with RouteConfigContext support
   const setLocation = useCallback(
     (to: string, options?: NavigateOptions) => {
+      const [pathPart, queryPart] = to.split("?");
```
[the remark this answers is not in the document]
Comment (kentwelcome) [reply js/src/lib/hooks/useAppRouter.ts]: Fixed! Applied the same `indexOf` approach with proper fragment handling here as well.

File: js/src/lib/hooks/useAppRouter.ts [hunk 4]
```diff
@@ -175,16 +202,42 @@ export function useAppNavigation() {
...
+        ? `${resolvedPath}?${queryPart}`
+        : resolvedPath;
+      router.push(fullPath, options);
+    },
+    [router, resolvePath],
+  );
+
+  const replace = useCallback(
+    (href: string, options?: { scroll?: boolean }) => {
+      const [pathPart, queryPart] = href.split("?");
```
[the remark this answers is not in the document]
Comment (kentwelcome) [reply js/src/lib/hooks/useAppRouter.ts]: Fixed! Applied the same `indexOf` approach with proper fragment handling.

File: js/src/lib/hooks/useAppRouter.ts [hunk 5]
```diff
@@ -175,16 +202,42 @@ export function useAppNavigation() {
...
+      const resolvedPath = resolvePath(pathPart);
+      const fullPath = queryPart
+        ? `${resolvedPath}?${queryPart}`
+        : resolvedPath;
+      router.replace(fullPath, options);
+    },
+    [router, resolvePath],
```
[the remark this answers is not in the document]
Comment (kentwelcome) [reply js/src/lib/hooks/useAppRouter.ts]: Fixed! Implemented a `buildResolvedHref` helper function that properly handles path, query string, and fragment separation using `indexOf`. Both `push` and `replace` now use this helper for consistent behavior.

File: js/src/lib/hooks/RouteConfigContext.tsx [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+        return path;
+      }
+
+      // Handle paths that already start with the basePath (avoid double-prefixing)
+      if (path.startsWith(basePath)) {
+        return path;
+      }
+
+      // Handle absolute URLs (http://, https://, etc.) - don't prefix
+      if (path.match(/^https?:\/\//)) {
```
[the remark this answers is not in the document]
Comment (kentwelcome) [reply js/src/lib/hooks/RouteConfigContext.tsx]: Fixed! Changed regex to `/^[a-z][a-z0-9+.-]*:/i` which matches any valid URL scheme according to RFC 3986. This now correctly handles mailto:, tel:, ftp:, data:, and any other valid URI schemes.

File: js/src/lib/hooks/useAppRouter.ts [hunk 7]
```diff
@@ -38,31 +45,44 @@ interface NavigateOptions {
...
       } else {
-        router.push(to, { scroll: options?.scroll ?? true });
+        router.push(fullPath, { scroll: options?.scroll ?? true });
       }
     },
-    [router],
+    [router, resolvePath],
   );
 
   return [pathname, setLocation];
```
[the remark this answers is not in the document]
Comment (kentwelcome) [reply js/src/lib/hooks/useAppRouter.ts:108]: Fixed! Added `stripBasePath` function to RouteConfigContext and applied it to all hooks that return pathname:
- `useAppLocation`
- `useAppLocationWithSearch`
- `useAppNavigation`
- `useAppRoute`

Now components receive logical paths (e.g., `/checks`) instead of physical paths (e.g., `/oss/abc123/checks`), maintaining backward compatibility with existing route matching code like `isCheckDetailPage(location)`.

**Comment by kentwelcome** [comment]
Close this PR, we already implemented the same logic in #1046 
We will also update our UI code structure to incorporate a new design.
